tsl/src/continuous_aggs/create.c

```diff
@@ -3469,13 +3554,35 @@ build_union_query(CAggTimebucketInfo *tbinfo, int matpartcolno, Query *q1, Query
 	 * If there is join in CAgg definition then adjust varno
 	 * to get time column from the hypertable in the join.
 	 */
-	if (list_length(q2->rtable) == CONTINUOUS_AGG_MAX_JOIN_RELATIONS)
+	if (list_length(q2->rtable) == CONTINUOUS_AGG_MAX_JOIN_RELATIONS ||
+		!IsA(linitial(q2->jointree->fromlist), RangeTblRef))
 	{
 		Oid normal_table_id = InvalidOid;
-		RangeTblRef *rtref = linitial_node(RangeTblRef, q2->jointree->fromlist);
-		RangeTblEntry *rte = list_nth(q2->rtable, rtref->rtindex - 1);
-		RangeTblRef *rtref_other = lsecond_node(RangeTblRef, q2->jointree->fromlist);
-		RangeTblEntry *rte_other = list_nth(q2->rtable, rtref_other->rtindex - 1);
+		RangeTblEntry *rte = NULL;
+		RangeTblEntry *rte_other = NULL;
+
+		if (list_length(q2->rtable) == CONTINUOUS_AGG_MAX_JOIN_RELATIONS)
+		{
+			RangeTblRef *rtref = linitial_node(RangeTblRef, q2->jointree->fromlist);
+			rte = list_nth(q2->rtable, rtref->rtindex - 1);
+			RangeTblRef *rtref_other = lsecond_node(RangeTblRef, q2->jointree->fromlist);
+			rte_other = list_nth(q2->rtable, rtref_other->rtindex - 1);
+		}
+		else if (!IsA(linitial(q2->jointree->fromlist), RangeTblRef))
```

Comment:
As far as I have understood the jointree, at leaves we have RangeTblRefs and above them is a JoinExpr (if any). So, if there is a JoinExpr it is always above the RangeTblRef nodes so it will appear in the first node otherwise not. Other case of joins, when no explicit JOIN is used in the query, then we decipher it by rtable entries.

Reply:
Could you add a comment explaining the situation?

Reply:
+1

Reply:
Added the comment for this.